3170 informed consent

pfriedman opened the pull request:
This fairly large change addresses issues with Informed Consent. Several of these changes are already being tested in various environments as at times the head of this branch has been deployed. Things being tested are Reconsent, Withdrawal, and Child Consent.

There are several conflicts between this branch and the current master and a rebase or patch will need to happen before this PR can even be considered to be merged.

Commit:
Renamed method title -> page_title

Commit:
Renamed method title -> page_title

Commit:
Renamed method title -> page_title

Commit:
Renamed method title -> page_title

Commit:
Do not override method #3526

This override was necessary at one point, iirc, to
handle how surveyor was displaying datetime fields
in the UI. Surveyor 1.2 updated this method and there
is no longer a need to override. In fact, date and time
fields in Surveyor are broken in Cases due to this override
so we should not have this at all.

Overridden methods should be rare.

Commit:
Do not override method #3526

This override was necessary at one point, iirc, to
handle how surveyor was displaying datetime fields
in the UI. Surveyor 1.2 updated this method and there
is no longer a need to override. In fact, date and time
fields in Surveyor are broken in Cases due to this override
so we should not have this at all.

Overridden methods should be rare.

Commit:
Do not override method #3526

This override was necessary at one point, iirc, to
handle how surveyor was displaying datetime fields
in the UI. Surveyor 1.2 updated this method and there
is no longer a need to override. In fact, date and time
fields in Surveyor are broken in Cases due to this override
so we should not have this at all.

Overridden methods should be rare.

Commit:
Redirect to contact_link decision page #3471

upon update redirect to the contact link decision page
rather than the edit contact link page

Commit:
Redirect to contact_link decision page #3471

upon update redirect to the contact link decision page
rather than the edit contact link page

Commit:
Redirect to contact_link decision page #3471

upon update redirect to the contact link decision page
rather than the edit contact link page

Commit:
extract method refactoring

Commit:
extract method refactoring

Commit:
extract method refactoring

Commit:
Action to begin administration of NIR #3471

This action (like the instrument start and participant_consent start actions)
initiates the administration of a Survey and associates it with the
non_interview_report mdes record.

Commit:
Action to begin administration of NIR #3471

This action (like the instrument start and participant_consent start actions)
initiates the administration of a Survey and associates it with the
non_interview_report mdes record.

Commit:
Action to begin administration of NIR #3471

This action (like the instrument start and participant_consent start actions)
initiates the administration of a Survey and associates it with the
non_interview_report mdes record.

Commit:
Added association and start! method #3471

Similar to the ParticipantConsent and Instrument (other
MDES record that are associated with a ResponseSet),
we add the association and a start! method to create
the relationship to a record and the ResponseSet associated
with the given Survey.

Commit:
Added association and start! method #3471

Similar to the ParticipantConsent and Instrument (other
MDES record that are associated with a ResponseSet),
we add the association and a start! method to create
the relationship to a record and the ResponseSet associated
with the given Survey.

Commit:
Added association and start! method #3471

Similar to the ParticipantConsent and Instrument (other
MDES record that are associated with a ResponseSet),
we add the association and a start! method to create
the relationship to a record and the ResponseSet associated
with the given Survey.

Commit:
Helper methods on ResponseSet #3260

With the addition of the ParticipantConsent and NonInterviewReport
MDES record associations on ResponseSet, we needed helper
methods to determine the contact_link or event associated with
the ResponseSet through the various MDES associations.

Commit:
Helper methods on ResponseSet #3260

With the addition of the ParticipantConsent and NonInterviewReport
MDES record associations on ResponseSet, we needed helper
methods to determine the contact_link or event associated with
the ResponseSet through the various MDES associations.

Commit:
Helper methods on ResponseSet #3260

With the addition of the ParticipantConsent and NonInterviewReport
MDES record associations on ResponseSet, we needed helper
methods to determine the contact_link or event associated with
the ResponseSet through the various MDES associations.

Commit:
Defer event to ResponseSet #3260

Commit:
Defer event to ResponseSet #3260

Commit:
Defer event to ResponseSet #3260

Commit:
NIR survey has only one part #3471

and as such is the final_survey_part

Commit:
NIR survey has only one part #3471

and as such is the final_survey_part

Commit:
NIR survey has only one part #3471

and as such is the final_survey_part

Commit:
First pass at form update #3559

Only show NIR sections based on whether or not
the NIR is associated with a ResponseSet (i.e. created from the
IRB_CON_NonInterviewReport).

Commit:
First pass at form update #3559

Only show NIR sections based on whether or not
the NIR is associated with a ResponseSet (i.e. created from the
IRB_CON_NonInterviewReport).

Commit:
First pass at form update #3559

Only show NIR sections based on whether or not
the NIR is associated with a ResponseSet (i.e. created from the
IRB_CON_NonInterviewReport).

Commit:
update method description

Commit:
update method description

Commit:
update method description

Commit:
update method description

Commit:
inline variable refactoring

Commit:
inline variable refactoring

Commit:
inline variable refactoring

Commit:
add routes for non_interview_report #3471

Commit:
add routes for non_interview_report #3471

Commit:
add routes for non_interview_report #3471

Commit:
Add has_many NIR association to person

Commit:
Add has_many NIR association to person

Commit:
Add has_many NIR association to person

Commit:
Non Interview Report Survey #3471

Created a Survey for the Non Interview Report and
corresponding OperationalDataExtractor.

Commit:
Non Interview Report Survey #3471

Created a Survey for the Non Interview Report and
corresponding OperationalDataExtractor.

Commit:
Non Interview Report Survey #3471

Created a Survey for the Non Interview Report and
corresponding OperationalDataExtractor.

Commit:
View changes and helper methods #3471

added to handle the addition of the Non Interview Report
as a Survey

Commit:
View changes and helper methods #3471

added to handle the addition of the Non Interview Report
as a Survey

Commit:
View changes and helper methods #3471

added to handle the addition of the Non Interview Report
as a Survey

Commit:
View changes and helper methods #3471

added to handle the addition of the Non Interview Report
as a Survey

Commit:
Create ParticipantConsentSamples #3534

when creating operational data from the responses in
the Informed Consent surveys.

Commit:
Create ParticipantConsentSamples #3534

when creating operational data from the responses in
the Informed Consent surveys.

Commit:
Create ParticipantConsentSamples #3534

when creating operational data from the responses in
the Informed Consent surveys.

Commit:
Create ParticipantConsentSamples #3534

when creating operational data from the responses in
the Informed Consent surveys.

Commit:
added idea of unambiguous code

Most NcsCodedAttributes across MDES records are uniquely
named (that the table column name used matches the list_name).
MDES 3.0 introduced the non_interview_report_provider which
shares the same column name as the non_interview_report but
references different code lists.

this commit alleviates that in a shared reused partial by introducing
an unambiguous_code variable for use in the partial.

Commit:
added idea of unambiguous code

Most NcsCodedAttributes across MDES records are uniquely
named (that the table column name used matches the list_name).
MDES 3.0 introduced the non_interview_report_provider which
shares the same column name as the non_interview_report but
references different code lists.

this commit alleviates that in a shared reused partial by introducing
an unambiguous_code variable for use in the partial.

Commit:
added idea of unambiguous code

Most NcsCodedAttributes across MDES records are uniquely
named (that the table column name used matches the list_name).
MDES 3.0 introduced the non_interview_report_provider which
shares the same column name as the non_interview_report but
references different code lists.

this commit alleviates that in a shared reused partial by introducing
an unambiguous_code variable for use in the partial.

Commit:
renamed method

to a more generic method name

Commit:
renamed method

to a more generic method name

Commit:
renamed method

to a more generic method name

Commit:
Refusal Non Interview Report #3471

Records associated with the Non Interview Report

Commit:
Refusal Non Interview Report #3471

Records associated with the Non Interview Report

Commit:
Refusal Non Interview Report #3471

Records associated with the Non Interview Report

Commit:
added missing param

Commit:
added missing param

Commit:
added missing param

Commit:
Update to PSC Template #3471

To handle Non Interview Report

Commit:
Update to PSC Template #3471

To handle Non Interview Report

Commit:
Update to PSC Template #3471

To handle Non Interview Report

Commit:
Update to PSC Template #3471

To handle Non Interview Report

Commit:
corrected ncs code list name

Commit:
Combine all internal surveys #3260

Use dependencies for each informed consent survey type
instead of separate surveys

Commit:
Combine all internal surveys #3260

Use dependencies for each informed consent survey type
instead of separate surveys

Commit:
Combine all internal surveys #3260

Use dependencies for each informed consent survey type
instead of separate surveys

Commit:
Combine all internal surveys #3260

Use dependencies for each informed consent survey type
instead of separate surveys

Commit:
informed consent #3170

items that were missed in previous commits
- added who withdrew consent to ODE
- reworded items in view to be more general
- use the one internal survey in specs

Commit:
informed consent #3170

items that were missed in previous commits
- added who withdrew consent to ODE
- reworded items in view to be more general
- use the one internal survey in specs

Commit:
informed consent #3170

items that were missed in previous commits
- added who withdrew consent to ODE
- reworded items in view to be more general
- use the one internal survey in specs

Commit:
informed consent #3170

items that were missed in previous commits
- added who withdrew consent to ODE
- reworded items in view to be more general
- use the one internal survey in specs

Commit:
temp. fix for issue w/ multiple children #3445

Until we address #3445 multiple birth was showing a
participant consent section for each child. This is a
/very/ temporary fix for demo purposes only.

Commit:
temp. fix for issue w/ multiple children #3445

Until we address #3445 multiple birth was showing a
participant consent section for each child. This is a
/very/ temporary fix for demo purposes only.

Commit:
temp. fix for issue w/ multiple children #3445

Until we address #3445 multiple birth was showing a
participant consent section for each child. This is a
/very/ temporary fix for demo purposes only.

Commit:
temp. fix for issue w/ multiple children #3445

Until we address #3445 multiple birth was showing a
participant consent section for each child. This is a
/very/ temporary fix for demo purposes only.

Commit:
Child Consent #3445

In allowing for multiple consents for a child participant
the way we create child participants needed to change.
Previously the child consent itself created the child participant record,
this commit de-couples the child person/participant record creation
from the child participant consent record adding a way to create/update
that child record on it's own.

Commit:
Child Consent #3445

In allowing for multiple consents for a child participant
the way we create child participants needed to change.
Previously the child consent itself created the child participant record,
this commit de-couples the child person/participant record creation
from the child participant consent record adding a way to create/update
that child record on it's own.

Commit:
Child Consent #3445

In allowing for multiple consents for a child participant
the way we create child participants needed to change.
Previously the child consent itself created the child participant record,
this commit de-couples the child person/participant record creation
from the child participant consent record adding a way to create/update
that child record on it's own.

Commit:
View and helper changes for #3445

The new child consent workflow required a fair number
of updates to the contact_links partials which show the
links of activities to perform.

Commit:
View and helper changes for #3445

The new child consent workflow required a fair number
of updates to the contact_links partials which show the
links of activities to perform.

Commit:
View and helper changes for #3445

The new child consent workflow required a fair number
of updates to the contact_links partials which show the
links of activities to perform.

Commit:
View and helper changes for #3445

The new child consent workflow required a fair number
of updates to the contact_links partials which show the
links of activities to perform.

Commit:
On demand Child consent #3445

Updated PSC Template to allow for scheduling of ad hoc
Child Consents. Still need to implement the consent action
in Cases, but this at least creates the segments for consent.

Also associated with #3171

Commit:
On demand Child consent #3445

Updated PSC Template to allow for scheduling of ad hoc
Child Consents. Still need to implement the consent action
in Cases, but this at least creates the segments for consent.

Also associated with #3171

Commit:
On demand Child consent #3445

Updated PSC Template to allow for scheduling of ad hoc
Child Consents. Still need to implement the consent action
in Cases, but this at least creates the segments for consent.

Also associated with #3171

Commit:
On demand Child consent #3445

Updated PSC Template to allow for scheduling of ad hoc
Child Consents. Still need to implement the consent action
in Cases, but this at least creates the segments for consent.

Also associated with #3171

Commit:
Move Child management to new partial #3445

Commit:
Move Child management to new partial #3445

Commit:
Move Child management to new partial #3445

Commit:
Refactor method to use SQL

to replace calls to .collect .uniq .flatten and .compact

Commit:
Refactor method to use SQL

to replace calls to .collect .uniq .flatten and .compact

Commit:
Refactor method to use SQL

to replace calls to .collect .uniq .flatten and .compact

Commit:
Various view changes #3445

to handle Child Consents

Commit:
Various view changes #3445

to handle Child Consents

Commit:
Various view changes #3445

to handle Child Consents

Commit:
Reformat relationships tab - wip

Commit:
Reformat relationships tab - wip

Commit:
Reformat relationships tab - wip

Commit:
Set attribute @relationship_code

Commit:
Set attribute @relationship_code

Commit:
Set attribute @relationship_code

Commit:
Ad-hoc child consent

Commit:
Ad-hoc child consent

Commit:
Ad-hoc child consent

Commit:
Revision 47 - PSC Template #3170

Rename the Informed Consent activities to more
accurate names

Commit:
Revision 47 - PSC Template #3170

Rename the Informed Consent activities to more
accurate names

Commit:
Revision 47 - PSC Template #3170

Rename the Informed Consent activities to more
accurate names

Commit:
Revision 47 - PSC Template #3170

Rename the Informed Consent activities to more
accurate names

Commit:
instrument type was incorrect

Consents are not associated with Instruments

Commit:
instrument type was incorrect

Consents are not associated with Instruments

Commit:
instrument type was incorrect

Consents are not associated with Instruments

Commit:
Update enroll status #3543

When a participant gives or withdraws consent

Commit:
Update enroll status #3543

When a participant gives or withdraws consent

Commit:
Update enroll status #3543

When a participant gives or withdraws consent

Commit:
Determine consent type #3170

Is this consent a child consent (or which child consent)
or is it a withdrawal or a reconsent?

Commit:
Determine consent type #3170

Is this consent a child consent (or which child consent)
or is it a withdrawal or a reconsent?

Commit:
Determine consent type #3170

Is this consent a child consent (or which child consent)
or is it a withdrawal or a reconsent?

Commit:
fixed merge conflict

Commit:
Add new contact link #3375

to allow users to start a contact for any event.
Other commits included as view cleanup when testing
this addition.

Closes #3375

Commit:
Add new contact link #3375

to allow users to start a contact for any event.
Other commits included as view cleanup when testing
this addition.

Closes #3375

Commit:
InvalidSurveyException

was being raised improperly

Commit:
InvalidSurveyException

was being raised improperly

Commit:
Added low to high event

Participant#advance takes an event (or the most recent event)
to determine which event to send to update_state_to_next_event

This method was missing the low to high event as a transition

Commit:
Added low to high event

Participant#advance takes an event (or the most recent event)
to determine which event to send to update_state_to_next_event

This method was missing the low to high event as a transition

Commit:
use form in psc labels #3659

for the consent surveys use "form:" instead of
"instrument:" to preface

Closes #3659

Commit:
use form in psc labels #3659

for the consent surveys use "form:" instead of
"instrument:" to preface

Closes #3659

Commit:
update deploy notes

Commit:
update deploy notes

Commit:
Multiple Child Consents #3445

Handle multiple child consents of all types
birth to 6 months and
6 months to age of majority

Commit:
Multiple Child Consents #3445

Handle multiple child consents of all types
birth to 6 months and
6 months to age of majority

Commit:
minor view updates

remove title of page (already titled in tab)
update conditions on consent form

Commit:
minor view updates

remove title of page (already titled in tab)
update conditions on consent form

Commit:
fixed merge conflicts for pull request

Commit:
Child Reconsent/Withdrawal #3445

Allow user to schedule Child Consent 6 Months to Age of Majority
at will for Reconsent or Withdrawal

Commit:
Create Withdrawal PPG Status #3574

If the Participant withdraws from the Study create a PpgStatusHistory record
with a ppg status of 'withdrawn'

Closes #3574

Commit:
Move adjudication to EventsController

from the SurveyorController as some Participants who are ineligible
do not take the Screener instrument

Closes #3734

Commit:
Return NIR form to event #3736

For all non-consent events allow user to create a
Non-Interview Report

Mostly reverts commit 90b86267f51ea

Closes #3736

Commit:
remove files that are no longer used

Commit:
update to determine if consented #3560

Use the most recent consent to determine if the
participant has consented

Commit:
Re-work logic for consented? #3560

For determining if the participant is consented?
reconsented? or withdrawn?

or consented to sample collection

we use the most recent ParticipantConsent record to
determine consent.

Missing from this commit is a test if the most recent
consent has sample consents and if this is not
true (and consent is given) to check the previous
consent's sample consent records.

Commit:
fixed whitespace

.haml got indented improperly

Commit:
added check for nil

Commit:
only use sortable consents

to sort the consents

jdzak commented on app/controllers/events_controller.rb:
```diff
@@ -50,6 +50,10 @@ def update
+        if @event.screener_event?
+          EligibilityAdjudicator.adjudicate_eligibility(@event.participant.person)
+        end
+
```
Just a reminder to move this to a place where Field can also benefit from this.

hannahwhy replied:
Agreed.

On a related note, I'd like to know why useful business logic keeps getting stuck in controllers.  (A lot of this `update` action seems like it really needs to be pushed to the model layer.)  I thought it was clear that such code can no longer be confined to a single request.

Commit:
Remove the ParticipantConsent controller #3626

In favor of always using the InformedConsent survey.
Legacy ParticipantConsent records can create an associated
response_set with which to use in Surveyor.
The edit action remains as a hook to determine whether or not
to create a response_set for a legacy consent.

Closes #3626

Commit:
Move eligibility adjudication #3734

to an after save hook on Event
only for screener events

Closes #3734

Commit:
Use Survey for Consent #3260 #3626

The removal of the Participant Consent Controller
requires that the creation of the associated Informed
Consent Event with another Event (e.g. PV1) be moved
to another location. The creation of the ParticipantConsent
seemed as logical a place as any.

Commit:
Fixed failing spec

to have expected be the actual what is expected

Commit:
Fixed failing specs

Commit:
updated data export identifiers #3260

on the Participant Consent survey.

pfriedman commented:
As much of this PR is already in testing on the QA servers and as it is needed for the 1.5.0 release candidate I am merging this PR

jdzak commented on vendor/ncs_navigator_schema:
```diff
@@ -1 +1 @@
-Subproject commit a2bcbe67bcb06fc32e93165e7cd45b95beede417
+Subproject commit c9b696297c7d857a492be9c04b55afeb65e0d7ce
```
Watch out here, this should be c9b696297c7d857a492be9c04b55afeb65e0d7ce.

jdzak replied:
oops, other way around.. a2bcbe67bcb06fc32e93165e7cd45b95beede417
